### PR TITLE
Nullable annotate System.IO.Pipes

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Contributes to https://github.com/dotnet/runtime/issues/2339

### CODE REVIEW

File: src/libraries/System.IO.Pipes/ref/System.IO.Pipes.cs [hunk 1]
```diff
@@ -111,16 +111,17 @@ public abstract partial class PipeStream : System.IO.Stream
         public virtual System.IO.Pipes.PipeTransmissionMode ReadMode { get { throw null; } set { } }
         public Microsoft.Win32.SafeHandles.SafePipeHandle SafePipeHandle { get { throw null; } }
         public virtual System.IO.Pipes.PipeTransmissionMode TransmissionMode { get { throw null; } }
-        public override System.IAsyncResult BeginRead(byte[] buffer, int offset, int count, System.AsyncCallback callback, object state) { throw null; }
-        public override System.IAsyncResult BeginWrite(byte[] buffer, int offset, int count, System.AsyncCallback callback, object state) { throw null; }
+        public override System.IAsyncResult BeginRead(byte[] buffer, int offset, int count, System.AsyncCallback callback, object? state) { throw null; }
+        public override System.IAsyncResult BeginWrite(byte[] buffer, int offset, int count, System.AsyncCallback callback, object? state) { throw null; }
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipes/ref/System.IO.Pipes.cs]: The `callback` should be nullable for both BeginXx methods, too.

File: src/libraries/System.IO.Pipes/ref/System.IO.Pipes.cs [hunk 2]
```diff
@@ -111,16 +111,17 @@ public abstract partial class PipeStream : System.IO.Stream
...
         protected internal virtual void CheckPipePropertyOperations() { }
         protected internal void CheckReadOperations() { }
         protected internal void CheckWriteOperations() { }
         protected override void Dispose(bool disposing) { }
         public override int EndRead(System.IAsyncResult asyncResult) { throw null; }
         public override void EndWrite(System.IAsyncResult asyncResult) { }
         public override void Flush() { }
-        protected void InitializeHandle(Microsoft.Win32.SafeHandles.SafePipeHandle handle, bool isExposed, bool isAsync) { }
+        public override System.Threading.Tasks.Task FlushAsync(System.Threading.CancellationToken cancellationToken) { throw null; }
+        protected void InitializeHandle(Microsoft.Win32.SafeHandles.SafePipeHandle? handle, bool isExposed, bool isAsync) { }
```
Comment (stephentoub) [inline src/libraries/System.IO.Pipes/ref/System.IO.Pipes.cs]: I think the `SafePipeHandle` should be non-nullable.

**Review by stephentoub — APPROVED** [review]
Other than the comments to be addressed, LGTM.

File: src/libraries/System.IO.Pipes/ref/System.IO.Pipes.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.IO.Pipes/ref/System.IO.Pipes.cs]: The System.IO.Stream annotation doesn't admit nullable callbacks:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L343
Should we revisit?

Comment (stephentoub) [reply src/libraries/System.IO.Pipes/ref/System.IO.Pipes.cs]: > Should we revisit?

Yes, that's a bug.

Comment (stephentoub) [reply src/libraries/System.IO.Pipes/ref/System.IO.Pipes.cs]: (I'll fix them.)

File: src/libraries/System.IO.Pipes/ref/System.IO.Pipes.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.IO.Pipes/ref/System.IO.Pipes.cs]: https://github.com/dotnet/runtime/blob/master/src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeServerStream.Unix.cs#L128

**Pull request merged**